keps/sig-node/20191211-node-maintenance-lease.md

```diff
@@ -0,0 +1,175 @@
+status: provisional
+see-also:
+  - "https://github.com/kubernetes/enhancements/issues/1403"
+---
+
+# Node Maintenance Lease
+
+## Table of Contents
+
+<!-- toc -->
+- [Release Signoff Checklist](#release-signoff-checklist)
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories [optional]](#user-stories-optional)
+    - [Story 1](#story-1)
+    - [Story 2](#story-2)
+  - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+    - [Impact to the Kubelet](#impact-to-the-kubelet)
+    - [Impact to Scheduleability](#impact-to-scheduleability)
+    - [Client Behavior](#client-behavior)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Alternatives](#alternatives)
+  - [Annotations on Node or elsewhere](#annotations-on-node-or-elsewhere)
+  - [Field in Node Spec Or Status](#field-in-node-spec-or-status)
+  - [Use existing system namespace for lease object](#use-existing-system-namespace-for-lease-object)
+  - [Use optional non-system namespace](#use-optional-non-system-namespace)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+
+Node Maintenance Leases are useful to coordinate disruptive actions
+controllers or administrators might wish to take against a node. This lease
+is a voluntary mechanism and allows coordination among different components
+without those components needing knowledge of each other.
+
+## Motivation
+
+Currently, there is no centralized way to inform a controller or user that
+another controller or user is performing a disruptive operation against a node.
+Examples of disruptive operations include:
+1. Rebooting a node
+1. Powering off a node for maintenance
+1. Draining or Cordoning a node
+1. Deleting/deprovisioning a node due to failed health checks
+
+### Goals
+
+1. Define a central point of coordination for various actors.
+1. Define high-level steps for well-behaved clients to implement.
+
+### Non-Goals
+
+Define actors which might utilize the lease.
+
+## Proposal
+
+Utilize the existing `Lease` built-in API in the API group `coordination.k8s.io`.
+Create a new Lease object per-node  with Name equal to Node name in a newly
```

Comment:
@derekwaynecarr I can't think of a use-case where a device plugin would ever need to perform an operation that was so disruptive to a node, that it would need to acquire one of these leases.

However, it might be useful for a device plugin to use this lease to discover that the node they are connected to is in maintenance mode. I can't think of a concrete use-case off the top of my head, but it seems likely that there is one.

Reply:
Possible scenario: a device plugin that activates a dongle or requests per-cpu-package licenses from a software license server.

Nodes in maintenance mode don't require dongle activations even if the node would be otherwise willing to mark itself as available for scheduling Pods that require a license.